ams/server/src/main/java/com/netease/arctic/server/optimizing/OptimizingQueue.java

```diff
@@ -330,75 +216,105 @@ public void updateOptimizerGroup(ResourceGroup optimizerGroup) {
+      long currentTime = System.currentTimeMillis();
+      scheduleTableIfNecessary(currentTime);
+      return waitDeadline > currentTime
+          && planningCompleted.await(waitDeadline - currentTime, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      LOG.error("Schedule table interrupted", e);
+      return false;
     } finally {
-      planLock.unlock();
+      scheduleLock.unlock();
     }
   }
 
-  private void planTasks() {
-    long startTime = System.currentTimeMillis();
-    List<TableRuntime> scheduledTables = schedulingPolicy.scheduleTables();
-    LOG.debug("Calculating and sorting tables by quota : {}", scheduledTables);
-
-    if (scheduledTables.size() <= 0) {
-      return;
+  private void scheduleTableIfNecessary(long startTime) {
+    if (planningTables.size() < maxPlanningParallelism) {
+      Set<TableRuntime> skipTables = new HashSet<>(planningTables);
+      plannedKeepingTables.entrySet().stream()
+          .filter(entry -> startTime - entry.getValue() < minPlanningInterval)
+          .filter(entry -> isOptimizingBlocked(entry.getKey()))
+          .map(Map.Entry::getKey)
+          .forEach(skipTables::add);
+      Optional.ofNullable(schedulingPolicy.scheduleTable(skipTables))
+          .ifPresent(tableRuntime -> triggerAsyncPlanning(tableRuntime, skipTables, startTime));
     }
-    List<TableIdentifier> plannedTables = Lists.newArrayList();
-    for (TableRuntime tableRuntime : scheduledTables) {
-      LOG.debug("Planning table {}", tableRuntime.getTableIdentifier());
-      try {
-        AmoroTable<?> table = tableManager.loadTable(tableRuntime.getTableIdentifier());
-        OptimizingPlanner planner =
-            new OptimizingPlanner(
-                tableRuntime.refresh(table),
-                (ArcticTable) table.originalTable(),
-                getAvailableCore());
-        if (tableRuntime.isBlocked(BlockableOperation.OPTIMIZE)) {
-          LOG.info("{} optimize is blocked, continue", tableRuntime.getTableIdentifier());
-          continue;
-        }
-        plannedTables.add(table.id());
-        if (planner.isNecessary()) {
-          TableOptimizingProcess optimizingProcess = new TableOptimizingProcess(planner);
-          LOG.info(
-              "{} after plan get {} tasks",
-              tableRuntime.getTableIdentifier(),
-              optimizingProcess.getTaskMap().size());
-          optimizingProcess.taskMap.values().forEach(taskQueue::offer);
-          break;
-        } else {
-          tableRuntime.cleanPendingInput();
-        }
-      } catch (Throwable e) {
-        LOG.error(tableRuntime.getTableIdentifier() + " plan failed, continue", e);
-      }
+  }
+
+  private void triggerAsyncPlanning(
+      TableRuntime tableRuntime, Set<TableRuntime> skipTables, long startTime) {
+    LOG.info("Trigger planning table {}", tableRuntime.getTableIdentifier());
+    planningTables.add(tableRuntime);
+    doPlanning(tableRuntime)
+        .whenComplete(
```

Comment:
You can return the Future
```
return future.whenComplete()
```
and in waitTask using
```
future.get(waitDeadline - currentTime, TimeUnit.MILLISECONDS)
```
to replace  planningCompleted.await(waitDeadline - currentTime, TimeUnit.MILLISECONDS)